File: 0875-koko-eating-bananas/0875-koko-eating-bananas.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        long int l=1,r=1000000000;
        while(l<r){
            long mid=l+(r-l)/2;
            int ans=0;
            for(int i=0;i<piles.size();i++){
                ans+=(piles[i]/mid)+(piles[i]%mid==0 ? 0 : 1);
            }
            if(ans<=h){
                r=mid;
            }
            else{
                l=mid+1;
            }
        }
        return l;
    }
};

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.java
class Solution {
    public long helper(int[] piles,int k){
        long total=0;
        for(int p:piles){
            total+=(p+k-1)/k;
        }
        return total;
    }
    public int minEatingSpeed(int[] piles, int h) {
        int n=piles.length;
        int l=1;
        int r=0;
        int ans=r;

        for(int num:piles){
            r=Math.max(r,num);
        }

        while(l<=r){
            int m=l+(r-l)/2;

            long time=helper(piles,m);

            if(time<=h){
                ans=m;
                r=m-1;
            }else{
                l=m+1;
            }
        }
        return ans;
    }
}